File: 2022/phase2_dress_rehearsal/byuvrx/src/frostlab-inekf-54148f8ac0c7/tests/Core/test_SE2.cpp
#include "gtest/gtest.h"
#include <Eigen/Core>
#include <unsupported/Eigen/MatrixFunctions>
#include <InEKF/Core>

#define EXPECT_MATRICES_EQ(M_actual, M_expected) \
  EXPECT_TRUE(M_actual.isApprox(M_expected, 1e-6)) << "  Actual:\n" << M_actual << "\nExpected:\n" << M_expected


TEST(SE2, BaseConstructor1){
    Eigen::Matrix4d state = Eigen::Matrix4d::Identity();
    Eigen::Matrix<double,7,7> sigma = Eigen::Matrix<double,7,7>::Identity();
    Eigen::Vector2d aug = Eigen::Vector2d::Ones();

    InEKF::SE2<2,2> x(state, sigma, aug);

    EXPECT_MATRICES_EQ(state, x());
    EXPECT_MATRICES_EQ(sigma, x.Cov());
    EXPECT_MATRICES_EQ(aug, x.Aug());
    EXPECT_TRUE(x.Uncertain());
}

TEST(SE2, BaseConstructor2){
    Eigen::MatrixXd state = Eigen::Matrix4d::Identity();
    Eigen::Matrix<double,6,6> sigma = Eigen::Matrix<double,6,6>::Identity();
    Eigen::Matrix<double,1,1> aug = Eigen::Matrix<double,1,1>::Ones();

    // make sure it gets mad if we pass the wrong sigma
    EXPECT_THROW(InEKF::SE2<Eigen::Dynamic> x(state, sigma);, std::range_error);
    // Or doesn't if we pass the right one
    EXPECT_NO_THROW(  (InEKF::SE2<Eigen::Dynamic,1>(state, sigma, aug)) );
    // dynamic aug also works
    Eigen::Matrix4d state2 = Eigen::Matrix4d::Identity();
    EXPECT_NO_THROW(  (InEKF::SE2<2,Eigen::Dynamic>(state2, sigma, aug)) );
}

TEST(SE2, TangentConstructor1){
    Eigen::Matrix<double,6,1> x;
    x << 0, 1, 2, 3, 4, 5;

    InEKF::SE2<2,1> state(x);
    EXPECT_MATRICES_EQ(state.R()(), Eigen::Matrix2d::Identity());
    EXPECT_EQ(state()(0,2), 1);
    EXPECT_EQ(state()(1,2), 2);
    EXPECT_EQ(state()(0,3), 3);
    EXPECT_EQ(state()(1,3), 4);
    EXPECT_EQ(state.Aug()(0), 5);
}

TEST(SE2, TangentConstructor2){
    Eigen::VectorXd x(6);
    x << 0, 1, 2, 3, 4, 5;

    InEKF::SE2<Eigen::Dynamic,1> state(x);
    EXPECT_MATRICES_EQ(state.R()(), Eigen::Matrix2d::Identity());
    EXPECT_EQ(state()(0,2), 1);
    EXPECT_EQ(state()(1,2), 2);
    EXPECT_EQ(state()(0,3), 3);
    EXPECT_EQ(state()(1,3), 4);
    EXPECT_EQ(state.Aug()(0), 5);

    InEKF::SE2<2,Eigen::Dynamic> state2(x);
    EXPECT_MATRICES_EQ(state.R()(), Eigen::Matrix2d::Identity());
    EXPECT_EQ(state2()(0,2), 1);
    EXPECT_EQ(state2()(1,2), 2);
    EXPECT_EQ(state2()(0,3), 3);
    EXPECT_EQ(state2()(1,3), 4);
    EXPECT_EQ(state2.Aug()(0), 5);

    // InEKF::SE2<Eigen::Dynamic,Eigen::Dynamic> state3(x);
    EXPECT_THROW( (InEKF::SE2<Eigen::Dynamic,Eigen::Dynamic>(x)), std::range_error);
}

TEST(SE2, PlainConstructor){
    InEKF::SE2<> x(0,1,2);
    EXPECT_MATRICES_EQ(x.R()(), Eigen::Matrix2d::Identity());
    EXPECT_EQ(x()(0,2), 1);
    EXPECT_EQ(x()(1,2), 2);
}

TEST(SE2, AddCol){
    InEKF::SE2<Eigen::Dynamic> x;
    EXPECT_MATRICES_EQ(x(), Eigen::Matrix3d::Identity());

    x.addCol(Eigen::Vector2d::Ones(2));

    EXPECT_EQ(x()(0,3), 1);
    EXPECT_EQ(x()(1,3), 1);

    // TODO: Test adding to Cov

    InEKF::SE2<> y;
    EXPECT_THROW( y.addCol(Eigen::Vector2d::Ones()), std::range_error);
}

TEST(SE2, AddAug){
    InEKF::SE2<1,Eigen::Dynamic> x;
    EXPECT_EQ(x.Aug().rows(), 0);

    x.addAug(2);

    EXPECT_EQ(x.Aug()(0), 2);

    // TODO: Test adding to Cov

    InEKF::SE2<> y;
    EXPECT_THROW( y.addAug(2), std::range_error);
}

TEST(SE2, Inverse){
    InEKF::SE2<> x(1,1,1);
    EXPECT_MATRICES_EQ(x().inverse(), x.inverse()());
}

TEST(SE2, Exp){
    Eigen::Matrix<double,6,1> x;
    x << 0, 1, 2, 3, 4, 5;
    
    InEKF::SE2<2,1> ours = InEKF::SE2<2,1>::Exp(x);
    Eigen::Matrix4d theirs = InEKF::SE2<2,1>::Wedge(x).exp(); 
    
    EXPECT_MATRICES_EQ(ours(), theirs);
    EXPECT_MATRICES_EQ(ours.Aug(), x.tail(1));
    EXPECT_THROW((InEKF::SE2<Eigen::Dynamic,2>::Exp(x)), std::range_error);
}

TEST(SE2, Log){
    Eigen::Vector3d xi;
    xi << .1, 2, 3;
    InEKF::SE2<> x = InEKF::SE2<>::Exp(xi);

    EXPECT_MATRICES_EQ(x.log(), xi);
}

TEST(SE2, Wedge){
    Eigen::Matrix<double,6,1> x;
    x << 1, 2, 3, 4, 5, 6;

    Eigen::Matrix4d ours = InEKF::SE2<2,1>::Wedge(x); 
    Eigen::Matrix4d theirs;
    theirs << 0, -1, 2, 4,
                1, 0, 3, 5,
                0, 0, 0, 0,
                0, 0, 0, 0;

    EXPECT_MATRICES_EQ(ours, theirs);
    EXPECT_THROW((InEKF::SE2<Eigen::Dynamic,2>::Wedge(x)), std::range_error);
}

TEST(SE2, Adjoint){
    InEKF::SE2<> x(1, 2, 3);
    Eigen::Matrix3d ad = x.Ad();

    EXPECT_MATRICES_EQ(ad.bottomRightCorner(2,2), x.R()());
    EXPECT_DOUBLE_EQ(ad(0,0), 1);
    EXPECT_DOUBLE_EQ(ad(1,0), 3);
    EXPECT_DOUBLE_EQ(ad(2,0), -2);
}

File: 2022/phase2_dress_rehearsal/byuvrx/src/frostlab-inekf-54148f8ac0c7/python/tests/Core/test_SE2.py
import numpy as np
from numpy.testing import assert_allclose
from scipy.linalg import expm, logm
from numpy.linalg import inv
from inekf import SE2
import pytest

def test_BaseConstructor1():
    state = np.eye(4)
    sigma = np.eye(7)
    aug = np.zeros(2)

    x = SE2[2,2](state, sigma, aug)

    assert_allclose(state, x.State)
    assert_allclose(sigma, x.Cov)
    assert_allclose(aug, x.Aug)

def test_BaseConstructor2():
    state = np.eye(4)
    sigma = np.eye(6)
    aug = np.zeros(1)

    with pytest.raises(Exception):
        x = SE2["D"](state, sigma)

    x = SE2["D",1](state, sigma, aug)
    x = SE2[2,"D"](state, sigma, aug)

def test_TangentConstructor1():
    x = np.arange(6)

    state = SE2[2,1](x)

    assert_allclose(state.R().State, np.eye(2))
    assert state.State[0,2] == 1
    assert state.State[1,2] == 2
    assert state.State[0,3] == 3
    assert state.State[1,3] == 4
    assert state.Aug[0] == 5

def test_TangentConstructor2():
    x = np.arange(6)

    state = SE2["D", 1](x)
    assert_allclose(state.R().State, np.eye(2))
    assert state.State[0,2] == 1
    assert state.State[1,2] == 2
    assert state.State[0,3] == 3
    assert state.State[1,3] == 4
    assert state.Aug[0] == 5

def test_PlainConstructor():
    x = SE2(0,1,2)
    assert_allclose(x.R().State, np.eye(2))
    assert x.State[0,2] == 1
    assert x.State[1,2] == 2

def test_AddCol():
    x = SE2["D",0]()
    assert_allclose(x.State, np.eye(3))

    x.addCol(np.ones(2))

    assert_allclose(x[1], np.ones(2))

def test_AddAug():
    x = SE2[1,"D"]()

    x.addAug(2)
    assert x.Aug[-1] == 2

    with pytest.raises(Exception):
        y = SE2()
        y.addAug(2)

def test_Inverse():
    x = SE2()
    assert_allclose(inv(x.State), x.inverse().State)

def test_Exp():
    x = np.arange(6)

    ours = SE2[2,1].Exp(x)
    theirs = expm( SE2[2,1].Wedge(x) )

    assert_allclose(theirs, ours.State)
    assert x[-1] == ours.Aug[0]
    with pytest.raises(Exception):
        SE2["D",2].Exp(x)

def test_Log():
    xi = np.array([.1, 2, 3])
    x = SE2.Exp(xi)

    assert_allclose(x.log(), xi)

def test_Wedge():
    x = np.arange(1,7)
    ours = SE2[2,1].Wedge(x)
    theirs =  np.array([[0, -1, 2, 4],
                        [1, 0, 3, 5],
                        [0, 0, 0, 0],
                        [0, 0, 0, 0]])
    
    assert_allclose(theirs, ours)

    with pytest.raises(Exception):
        SE2["D",2].Wedge(x)

def test_Adjoint():
    x = SE2(1, 2, 3)
    ad = x.Ad()

    assert_allclose(ad[-2:,-2:], x.R().State)
    assert ad[0,0] == 1
    assert ad[1,0] == 3
    assert ad[2,0] == -2
